### DISCUSSION TITLE
Sql script files are considered binary

### DISCUSSION THREAD

**Discussion by Magnus-Ekstrom** [community/community#102633] [Pull Requests] [answered]
### Select Topic Area

Bug

### Body

When trying to review a PR that contains ".sql" files they are categorized as binary files. This does not seem to be correct. See the supplied image below.
".sql" is a common extension for sql-script files. In my case this is a script with initial data for an application and is not containing any binary data at all. 

Any ideas on why it behaves like this?

![sqlpull2](https://github.com/community/community/assets/90033704/038046cd-fb89-4e37-bfbe-3d1ec34d3e55)

**Comment by humbertodias** [comment]
Weird. I have created a small PR and the sql has been show.
![image](https://github.com/community/community/assets/9255997/615558e4-6c05-41b8-a94a-68a4017ec894)
https://github.com/humbertodias/docker-pygame/pull/1/files
Maybe the issue would be the file size

**Reply by humbertodias** [reply]
I pushed some extra lines and the message appears "Large diffs are not rendered by default."
<img width="1048" alt="image" src="https://github.com/community/community/assets/9255997/1de6fd9b-8e74-40d9-a212-4d19e6dbcb45">
Well, that's different from yours message.

**Comment by spenserblack** [comment]
I'm 99% sure that GitHub uses [linguist](https://github.com/github-linguist/linguist) to determine if a file is binary ([docs](https://github.com/github-linguist/linguist/blob/559a6426942abcae16b6d6b328147476432bf6cb/docs/how-linguist-works.md)). Here's the relevant code for determining [if a file is binary](https://github.com/github-linguist/linguist/blob/559a6426942abcae16b6d6b328147476432bf6cb/lib/linguist/blob_helper.rb#L125-L142) and [if a file is likely binary](https://github.com/github-linguist/linguist/blob/559a6426942abcae16b6d6b328147476432bf6cb/lib/linguist/blob_helper.rb#L64-L66). Does anything stand out as obvious? An unusual encoding, or is the file particularly large?

**Reply by Magnus-Ekstrom** [reply]
Interesting. 
I can not find any special encoding or special characters in my file though. However, it is 331KB, so it is quite large I suppose.
But large files usually gets flagged as just large and you can get the diff by manually expanding the diff.
Any ideas on what to look for in the file that makes GH think that it is a binary?

**Reply by spenserblack** [reply]
> However, it is 331KB, so it is quite large I suppose.
But large files usually gets flagged as just large and you can get the diff by manually expanding the diff.

:thinking: I don't think that's the issue then, since IIRC linguist's limit was 1 MB. I happened to do some digging for one of my own projects and stumbled on [this code](https://github.com/brianmario/charlock_holmes/blob/3f4f57cfbee6dcd11941f54f9d4da0ebc1402421/ext/charlock_holmes/encoding_detector.c#L61-L130) from Charlock Holmes (which linguist uses to detect a file's encoding and if it's binary). I'm not too familiar with C, but based on the code comment near the bottom of the selected lines it looks like a file is considered binary if there is a `NULL` byte in the file.

**Comment by Magnus-Ekstrom** [comment]
Ok, checked to see if there was any null etc, but could not find any. 
There are a lot of Unicode characters tough since these files contains some translations. There are cyrillic characters as well as most of the Asian language signs involved. 

I noticed that it also seems to depend on which view in Github am I am.

1. :heavy_check_mark: By just navigating to the file from the "Code" tab i Github i can see the file just fine with all characters displayed correctly. `github.com/repo/main/Deploy-Locally/Sql/file.sql`
2. :x: But if I click on "History" for that sql file and then click the last commit to see the most recent change it will tell me 'Binary file not shown' `github.com/repo/commit/b9a42bbca0000000000110008e0000000005c800`
3. :x: Also when looking at the diff in a PR I will the the 'Binary file not shown' `github.com/repo/pull/242/files`

So the bottom line is that I can not diff these files on GH. This is most frustrating when looking at PRs.

**Reply by spenserblack** [reply, marked as answer]
Thanks for researching this, it's been pretty educational! I can't say for sure, but it seems like diffs are treated differently than the actual files, then. Is a local git client able to generate diffs for these files (`git show b9a42bbca0000000000110008e0000000005c800`)? I wonder if GitHub is relying on Git itself for binary detection (at least for diffs).

**Comment by Magnus-Ekstrom** [comment]
I think you are on to something here. I do get the same behaviour when showing the diff locally with `git show`. 
``` 
diff --git a/Deploy-Locally/Sql/file.sql b/Deploy-Locally/Sql/file.sql
index dd00000..0000053 100644
Binary files a/Deploy-Locally/Sql/file.sql and b/Deploy-Locally/Sql/file.sql differ
```
Using the following git version.
```
git --version
git version 2.43.0.windows.1
```

I guess that is just how git works then, and not GitHub related. Good to have that sorted out. 👍 

If I get the time tomorrow I might create a sample file from the original and try to binary reduce it until I see a normal diff. But it is most likely related to the the amount of Unicode characters in the file. And since I can not remove all these characters I have to live with it. But it is good to know whats causing it.

Thanks for your time and help to investigate!

**Reply by Magnus-Ekstrom** [reply]
Verified!
When opening the file in Notepad++ and checking the encoding I can see that the file that is considered binary by git has the encoding `UTF16 LE BOM`. It does not matter if remove all its content. It is still considered binary. Changing it to `UTF-8` and everything works as expected.

**Reply by spenserblack** [reply]
No problem! You might want to mark the files as text in a `.gitattributes` file. I've mostly used `GLOB binary`, but you'll want to do the exact opposite :laughing:
I think this would do it:
```gitattributes
# .gitattributes
# NOTE: `binary` is a shortcut for `-diff -merge -text`
# I don't think the linguist-language is necessary, but there's no harm in adding it if you're already setting attributes for SQL files.
*.sql diff merge text linguist-language=SQL
```

[More reading on `.gitattributes`](https://git-scm.com/docs/gitattributes).

Coincidentally I happened to stumble on a comment somewhere that said that git reads the first 8000 bytes to determine if something is binary or not. So a hacky solution might be to add 8000 text-like bytes (like a header comment) to your files :laughing:
